### ISSUE TITLE
Improve (and fix?) Node3D/Spatial Transform Functionality

### ISSUE THREAD

**Issue by mrtripie** [godotengine/godot-proposals#2685]
### Describe the project you are working on

3D level block out addon, 2D RPG
### Describe the problem or limitation you are having in your project
I was finding that some functions that I expected to be there (such as being able to directly set the global rotation and scale) were missing, and generally found that looking at the documentation was a lot more confusing than I expected, considering I've made 3D games in other popular engines before. I decided to write a proposal for a few improvements, but I ended up finding more things I found confusing over it. Note that I'm likely mistaken on some of this.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There doesn't seem to be any obvious direct way to set the global translation/rotation and scale in Spatial/Node3D.

The difference between local and object-local space is really confusing:
- In the 3d Editor, if you enable the toolbar option Local Space, the gizmos become relative to the node. (same as Blender)
- In the Spatial class reference it says: "Affine operations (rotate, scale, translate) happen in parent's local coordinate system, unless the Spatial object is set as top-level. Affine operations in this coordinate system correspond to direct affine operations on the Spatial's transform. The word local below refers to this coordinate system. The coordinate system that is attached to the Spatial object itself is referred to as object-local coordinate system."
- So according to that:
  - Local space is relative to the parent (not matching the toolbar option)
  - Object-Local Space
- So I think according to that the properties translation, rotation, and rotation degrees are Local Space but scale is actually Object-Local Space? But all are referred to in the documentation as Local Space.
- The functions translate/rotate/scale and translate/rotate/scale_object_locale are confusing:
  - translate and translate_object_local have some issues:
    - The documentation for both are the same up until these parts: 
      - In the documentation for translate it says this: "Note that the translation offset is affected by the node's scale, so if scaled by e.g. (10, 1, 1), a translation by an offset of (2, 0, 0) would actually add 20 (2 * 10) to the X coordinate."
      - For translate_object_local it says this: "...in local space."
    - First, in my tests both do the same thing with the node's scale (rotation also effects both), so I think these functions give the exact same result.
    - Second, I think translate being affected by the node's own scale and rotation is a bug, as Local Space is supposed to be relative to the parent based on the documentation.
    - Third, translate_object_local should say its in Object-Local Space, not Local Space in the documentation, and have the note about the scale being added to it, as well as one for rotation (though I think its working correctly)
  - A scale() (Local Space) function exists in source, but is not exposed, I assume because it causes confusion since if you scaled a node that was rotated relative to its parent (Local Space), it would be skewed, similar to globally scaling a rotated object. Also it has the same name as the scale property. If we cleared this up it could be exposed, although its probably not that useful
  - rotate_object_local should make clear in the documentation that the resulting rotation will be different depending on the node's rotation property and if the scale is non-uniform, it will be skewed (is the scale skewing a bug? should't this happen with rotate instead? Or am I getting it mixed up this time?)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I'm not entirely sure how to fix this, but maybe something like this would work: Local Space could be renamed Parent Space, and Object-Local Space renamed Local Space meaning:
- The new Local Space matches with the Local Space toolbar option 
- translate/rotate/scale functions should be renamed to something like translate/rotate/scale_relative_to_parent
- The unexposed scale function could then be exposed as scale_relative_to_parent
- translate/rotate/scale_object_local should be rename to translate/rotate/scale_local
- global_translate/rotate/scale should be renamed to translate/rotate/scale_global to match with the others
- I think in the documentation/tooltip the position and rotation properties should say Parent Space, while scale should say Local Space
- Alternatively Object-Local space could be renamed Object Space

I think it would be helpful to add global_translation/rotation/rotation_degrees/scale properties to match Node2D.

These aren't that important but perhaps some other renaming if its okay to break backwards compatibility here: 
_ set_identity() seems unclear to someone not used to it, should it be renamed something like clear_transform()?
- Maybe change the name for position on Node2D or translation on Node3D to match each other (I prefer position as its shorter and could be easier to tell apart from translate functions and transform at a glance, although maybe translation would be preferred to more obviously match with translate functions)

I have a test scene in Godot 3.3 that can help confirming some of these issues. I also tested the translation issue in a Godot 4.0 build from 5 days ago and it had the same issue.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Yes, after some confusion, may be hard with some things.
### Is there a reason why this should be core and not an add-on in the asset library?
It is core.

**Comment by mrtripie** [comment]
Test scene: https://drive.google.com/file/d/1GP-a3mlhcUUMNlkw44lGs4HjUNGbHfbL/view?usp=sharing

**Comment by Zireael07** [comment]
I think local space tooltip is short for object local space, but yes I agree this is extremely confusing. At least translate_local works as I expect it to.

**Comment by aaronfranke** [comment]
> Second, I think translate being affected by the node's own scale and rotation is a bug, as Local Space is supposed to be relative to the parent based on the documentation.

If you want to translate relative to the parent, you can just do `transform.origin += your_vector` or use the properties on Node2D (`position += your_vector`) and Spatial/Node3D (`translation += your_vector`). No method is required.

**Comment by mrtripie** [comment]
> > Second, I think translate being affected by the node's own scale and rotation is a bug, as Local Space is supposed to be relative to the parent based on the documentation.
> 
> If you want to translate relative to the parent, you can just do `transform.origin += your_vector` or use the properties on Node2D (`position += your_vector`) and Spatial/Node3D (`translation += your_vector`). No method is required.

That's already how I do it, though that's not really the point. The point is that the difference between the 2 different types of local transform is really confusing, I think even the developer of these functions got them mixed up, as the two translate functions appear to do the same thing, when they shouldn't. And I think the people who made the documentation also got them mixed up because of the confusion. If the function exists, it should do what it's supposed to do.

Right now using these functions feels like guesswork. I think that the names could be changed to be clearer, although that will break projects, so maybe not as that will only lead to another round of confusion.

Do you think I should make an issue on the main repository about improving the documentation for these functions, as they don't fully explain the difference, and I think some are wrong?

And as for the two translate functions appearing to be the same, perhaps another issue about that, where if its confirmed either:

- fix translate to be relative to the parent as it should be, at the cost of some projects not working right and confusing people
- or as translate isn't really needed, deprecate it, tell people to replace their uses with translate_object_local, or use the simple tranlation + vector way to do it relative to the parent like it was supposed to be. 

Technically translate_object_local's source has an extra calculation in it than translate, which my testing is saying does nothing and could be erased, though maybe there is sometimes a difference in their results I don't know about?

**Comment by bluenote10** [comment]
@mrtripie I fully agree with you that this is very confusing, see the related proposal https://github.com/godotengine/godot-proposals/issues/1336
